File: desktop/src/shared/pet-outfits.js
export const PET_OUTFIT_SLOTS = ['head', 'neck', 'side', 'scene']

const PET_OUTFIT_CATALOG = {
  pig: [
    {
      id: 'pig_basic_scarf',
      slot: 'neck',
      unlockLevel: 1,
      labels: {
        'zh-CN': '基础小围巾',
        en: 'Everyday scarf',
      },
    },
    {
      id: 'pig_sleep_cap',
      slot: 'head',
      unlockLevel: 2,
      labels: {
        'zh-CN': '软绵睡帽',
        en: 'Sleep cap',
      },
    },
    {
      id: 'pig_bell',
      slot: 'side',
      unlockLevel: 3,
      labels: {
        'zh-CN': '提醒铃铛',
        en: 'Reminder bell',
      },
    },
    {
      id: 'pig_work_badge',
      slot: 'side',
      unlockLevel: 4,
      labels: {
        'zh-CN': '搭档工作牌',
        en: 'Partner badge',
      },
    },
    {
      id: 'pig_star_hat',
      slot: 'head',
      unlockLevel: 5,
      labels: {
        'zh-CN': '星星小帽',
        en: 'Star hat',
      },
    },
  ],
  cat: [],
  dog: [],
}

const SLOT_LABELS = {
  head: {
    'zh-CN': '头部',
    en: 'Head',
  },
  neck: {
    'zh-CN': '颈部',
    en: 'Neck',
  },
  side: {
    'zh-CN': '侧边道具',
    en: 'Side prop',
  },
  scene: {
    'zh-CN': '场景道具',
    en: 'Scene prop',
  },
}

function getLocale(language) {
  return language === 'zh-CN' ? 'zh-CN' : 'en'
}

export function getPetOutfitCatalog(petType, language = 'zh-CN') {
  const locale = getLocale(language)
  return (PET_OUTFIT_CATALOG[petType] || []).map((item) => ({
    ...item,
    label: item.labels[locale],
  }))
}

export function getPetOutfitItem(petType, itemId, language = 'zh-CN') {
  return getPetOutfitCatalog(petType, language).find((item) => item.id === itemId) || null
}

export function getPetOutfitSlotLabel(slot, language = 'zh-CN') {
  const labels = SLOT_LABELS[slot] || SLOT_LABELS.scene
  return labels[getLocale(language)]
}

export function getUnlockedPetOutfitIds(petType, level) {
  return getPetOutfitCatalog(petType)
    .filter((item) => item.unlockLevel <= level)
    .map((item) => item.id)
}

export function normalizePetOutfitState(value, petType, level) {
  const catalog = getPetOutfitCatalog(petType)
  const knownItems = new Map(catalog.map((item) => [item.id, item]))
  const unlockedIds = new Set(
    Array.isArray(value?.unlocked_outfit_ids)
      ? value.unlocked_outfit_ids.filter((itemId) => knownItems.has(itemId))
      : [],
  )
  getUnlockedPetOutfitIds(petType, level).forEach((itemId) => unlockedIds.add(itemId))

  const equippedOutfits = {}
  for (const [slot, itemId] of Object.entries(value?.equipped_outfits || {})) {
    const item = knownItems.get(itemId)
    if (
      PET_OUTFIT_SLOTS.includes(slot)
      && item?.slot === slot
      && unlockedIds.has(itemId)
    ) {
      equippedOutfits[slot] = itemId
    }
  }

  return {
    unlocked_outfit_ids: catalog
      .filter((item) => unlockedIds.has(item.id))
      .map((item) => item.id),
    equipped_outfits: equippedOutfits,
  }
}

export function applyPetOutfitPreview(outfit, petType, itemId) {
  const item = getPetOutfitItem(petType, itemId)
  const unlockedIds = Array.isArray(outfit?.unlocked_outfit_ids)
    ? [...outfit.unlocked_outfit_ids]
    : []
  const equippedOutfits = { ...(outfit?.equipped_outfits || {}) }

  if (!item || !unlockedIds.includes(item.id)) {
    return {
      unlocked_outfit_ids: unlockedIds,
      equipped_outfits: equippedOutfits,
    }
  }

  return {
    unlocked_outfit_ids: unlockedIds,
    equipped_outfits: {
      ...equippedOutfits,
      [item.slot]: item.id,
    },
  }
}

File: backend/app/services/pet_outfits.py
OUTFIT_SLOTS = ("head", "neck", "side", "scene")

PET_OUTFIT_CATALOG = {
    "pig": (
        {
            "id": "pig_basic_scarf",
            "slot": "neck",
            "unlock_level": 1,
        },
        {
            "id": "pig_sleep_cap",
            "slot": "head",
            "unlock_level": 2,
        },
        {
            "id": "pig_bell",
            "slot": "side",
            "unlock_level": 3,
        },
        {
            "id": "pig_work_badge",
            "slot": "side",
            "unlock_level": 4,
        },
        {
            "id": "pig_star_hat",
            "slot": "head",
            "unlock_level": 5,
        },
    ),
    "cat": (),
    "dog": (),
}


def get_pet_outfit_catalog(pet_type: str) -> tuple[dict, ...]:
    return PET_OUTFIT_CATALOG.get(pet_type, ())


def get_pet_outfit_item(pet_type: str, item_id: str) -> dict | None:
    return next(
        (item for item in get_pet_outfit_catalog(pet_type) if item["id"] == item_id),
        None,
    )


def merge_unlocked_outfit_ids(
    pet_type: str,
    level: int,
    current_ids: list[str] | None = None,
) -> list[str]:
    catalog = get_pet_outfit_catalog(pet_type)
    known_ids = {item["id"] for item in catalog}
    unlocked_ids = {
        item_id
        for item_id in (current_ids or [])
        if item_id in known_ids
    }
    unlocked_ids.update(
        item["id"]
        for item in catalog
        if item["unlock_level"] <= level
    )
    return [item["id"] for item in catalog if item["id"] in unlocked_ids]


def normalize_equipped_outfits(
    pet_type: str,
    unlocked_ids: list[str],
    equipped_outfits: dict[str, str] | None,
) -> dict[str, str]:
    unlocked = set(unlocked_ids)
    normalized = {}
    for slot, item_id in (equipped_outfits or {}).items():
        item = get_pet_outfit_item(pet_type, item_id)
        if (
            slot in OUTFIT_SLOTS
            and item is not None
            and item["slot"] == slot
            and item_id in unlocked
        ):
            normalized[slot] = item_id
    return normalized


def build_pet_outfit_state(
    pet_type: str,
    level: int,
    unlocked_ids: list[str] | None,
    equipped_outfits: dict[str, str] | None,
) -> dict:
    normalized_unlocked = merge_unlocked_outfit_ids(
        pet_type,
        level,
        unlocked_ids,
    )
    return {
        "unlocked_outfit_ids": normalized_unlocked,
        "equipped_outfits": normalize_equipped_outfits(
            pet_type,
            normalized_unlocked,
            equipped_outfits,
        ),
    }
